opencensus/exporters/trace/zipkin/zipkin_exporter.cc

```diff
@@ -0,0 +1,342 @@
+        // Failed to enable follow location.
+        error = curl_easy_strerror(res);
+      } else if (CURLE_OK != (res = curl_easy_setopt(curl, CURLOPT_MAXREDIRS,
+                                                     max_redirect_times_))) {
+        // Failed to set max redirect times.
+        error = curl_easy_strerror(res);
+      }
+    }
+
+    // Sending HTTP request to url.
+    if (CURLE_OK != (res = curl_easy_perform(curl))) {
+      // Failed to send http request.
+      error = curl_easy_strerror(res);
+    } else {
+      long status_code = 0;
+      double total_time = 0, uploaded_bytes = 0, upload_speed = 0;
+
+      if (CURLE_OK != (res = curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE,
+                                               &status_code))) {
+        // Failed to get status code.
+        error = curl_easy_strerror(res);
+      } else if (CURLE_OK != (res = curl_easy_getinfo(curl, CURLINFO_TOTAL_TIME,
+                                                      &total_time))) {
+        // Failed to get total time.
+        error = curl_easy_strerror(res);
+      } else if (CURLE_OK !=
+                 (res = curl_easy_getinfo(curl, CURLINFO_SIZE_UPLOAD,
+                                          &uploaded_bytes))) {
+        // Failed to get uploaded bytes.
+        error = curl_easy_strerror(res);
+      } else if (CURLE_OK !=
+                 (res = curl_easy_getinfo(curl, CURLINFO_SPEED_UPLOAD,
+                                          &upload_speed))) {
+        // Failed to get upload speed.
+        error = curl_easy_strerror(res);
+      } else {
+        // LOG(INFO) << "HTTP request finished, status " << status_code
+        //           << ", uploaded " << uploaded_bytes << " bytes in "
+        //           << total_time << " seconds (" << (upload_speed / 1024)
+        //           << " KB/s)";
+      }
+    }
+  }
+
+  // TODO: Log error if one exists.
+  curl_slist_free_all(headers);
+  curl_easy_cleanup(curl);
+}
+
+void ZipkinExporter::Register(const ZipkinExporterOptions &options) {
+  // Create new exporter.
+  static CurlEnv *curl_lib = new CurlEnv();
+  ZipkinExporter *exporter = new ZipkinExporter(options);
+
+  // Create client handler that will send encoded span information to
+  // collection server.
+  HttpClientOptions http_client_options(exporter->options_.url);
+  exporter->trace_client_ = std::unique_ptr<ExportClient>(
+      dynamic_cast<ExportClient *>(new HttpClient(http_client_options)));
+
+  ::opencensus::trace::exporter::SpanExporter::RegisterHandler(
+      absl::WrapUnique<::opencensus::trace::exporter::SpanExporter::Handler>(
+          exporter));
+}
+
+void ZipkinExporter::Export(
+    const std::vector<::opencensus::trace::exporter::SpanData> &spans) {
+  if (!spans.empty()) {
+    std::string msg = message_codec_->Encode(spans);
+    trace_client_->SendMessage(msg, msg.size());
+  }
+}
+
+void ZipkinExporter::ExportForTesting(
```

Comment:
IMO expose Encode() and unit-test that instead.

Reply:
ExportForTesting directly calls encode now and prints the returned string.

Reply:
Instead of printing the encoded string:
- Remove ExportForTesting entirely.
- Expose Encode() for testing.
- Construct a span, encode it, test that the encoded JSON equals what we expect.